dm/dm/master/shardddl/optimist.go

```diff
@@ -134,31 +139,158 @@ func (o *Optimist) ShowLocks(task string, sources []string) []*pb.DDLLock {
+						lockUnsynced = append(lockUnsynced, fmt.Sprintf("%s-%s", source, dbutil.TableName(schema, table)))
 					}
 				}
 			}
 		}
-		sort.Strings(l.Synced)
-		sort.Strings(l.Unsynced)
-		ret = append(ret, l)
+		sort.Strings(lockSynced)
+		sort.Strings(lockUnsynced)
+
+		if len(owners) == 0 {
+			owners = append(owners, "")
+			ddlGroups = append(ddlGroups, nil)
+		}
+		for i, owner := range owners {
+			ret = append(ret, &pb.DDLLock{
+				ID:       lock.ID,
+				Task:     lock.Task,
+				Mode:     config.ShardOptimistic,
+				Owner:    owner,
+				DDLs:     ddlGroups[i],
+				Synced:   lockSynced,
+				Unsynced: lockUnsynced,
+			})
+		}
 	}
-	return ret
+	return ret, err
+}
+
+// UnlockLock unlocks a shard DDL lock manually only when using `unlock-ddl-lock` command.
+// ID: the shard DDL lock ID.
+// source, upstreamSchema, upstreamTable: reveal the upstream table's info which we need to skip/exec
+// action: whether to skip/exec the blocking DDLs for the specified upstream table
+// NOTE: this function has side effects, if it failed, some status can't revert anymore.
+// NOTE: this function should not be called if the lock is still in automatic resolving.
+func (o *Optimist) UnlockLock(ctx context.Context, id, source, upstreamSchema, upstreamTable string, action pb.UnlockDDLLockOp) error {
+	o.mu.Lock()
+	defer o.mu.Unlock()
+	if o.closed {
+		return terror.ErrMasterOptimistNotStarted.Generate()
+	}
+	task := utils.ExtractTaskFromLockID(id)
+	// 1. find the lock.
+	lock := o.lk.FindLock(id)
+	if lock == nil {
+		return terror.ErrMasterLockNotFound.Generate(id)
+	}
+
+	// 2. check whether has resolved before (this often should not happen).
+	if lock.IsResolved() {
+		_, err := o.removeLock(lock)
+		return err
+	}
+
+	// 3. find out related info & operation
+	infos, ops, _, err := optimism.GetInfosOperationsByTask(o.cli, task)
+	if err != nil {
+		return terror.ErrMasterLockIsResolving.Generatef("fail to get info and operation for task %s", task)
+	}
+	l := 0
+	for i, info := range infos {
+		if info.Task == task && info.Source == source && info.UpSchema == upstreamSchema && info.UpTable == upstreamTable {
+			infos[l] = infos[i]
+			l++
+		}
+	}
+	if l != 1 {
+		return terror.ErrMasterLockIsResolving.Generatef("fail to find related info for lock %s", id)
+	}
+	infos = infos[:l]
+
+	l = 0
+	for j, op := range ops {
+		if op.Task == task && op.Source == source && op.UpSchema == upstreamSchema && op.UpTable == upstreamTable {
+			if op.ConflictStage != optimism.ConflictDetected {
+				return terror.ErrMasterLockIsResolving.Generatef("lock %s is in %s status, not conflicted", id, op.ConflictStage)
+			}
+			ops[l] = ops[j]
+			l++
+		}
+	}
+	if l != 1 {
+		return terror.ErrMasterLockIsResolving.Generatef("fail to find related operation for lock %s", id)
+	}
+	ops = ops[:l]
+
+	// 4. rewrite operation.DDLs to skip/exec DDLs
+	switch action {
+	case pb.UnlockDDLLockOp_ExecLock:
+		ops[0].DDLs = infos[0].DDLs
```

Comment:
If a info with multiple DDLs, this may cause some problems.
e.g.
```
alter table drop column a, xxx(which cause the error)
```
execute theses DDLs will drop a in downstream. But I think is OK not to deal with it